clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java

```diff
@@ -2279,6 +2279,45 @@ public void testCommitOffsetRequestAsyncAlwaysReceiveFencedException() {
                 coordinator.commitOffsetsSync(singletonMap(t1p, new OffsetAndMetadata(100L)), time.timer(Long.MAX_VALUE)));
     }
 
+    @Test
+    public void testConsumerRejoinAfterRebalance() throws Exception {
+        try (ConsumerCoordinator coordinator = prepareCoordinatorForCloseTest(true, false, Optional.of("group-id"))) {
+            coordinator.ensureActiveGroup();
+
+            prepareOffsetCommitRequest(singletonMap(t1p, 100L), Errors.REBALANCE_IN_PROGRESS);
+
+            assertThrows(CommitFailedException.class, () -> coordinator.commitOffsetsSync(
+                singletonMap(t1p, new OffsetAndMetadata(100L)),
+                time.timer(Long.MAX_VALUE)));
+
+            assertFalse(client.hasPendingResponses());
+            assertFalse(client.hasInFlightRequests());
+
+            client.prepareResponse(joinGroupFollowerResponse(1, "consumer", "leader", Errors.NONE));
+
+            MockTime time = new MockTime(1);
+
+            //onJoinPrepare will be executed and onJoinComplete will not.
+            boolean res = coordinator.joinGroupIfNeeded(time.timer(1));
+
+            assertFalse(res);
+            assertFalse(client.hasPendingResponses());
+            assertFalse(client.hasInFlightRequests());
```

Comment:
Hmm, why `client.hasInFlightRequests()` is false? I thought it should have the sync-group request queued up here? Could you explain just for my own education.

Reply:
It's a very good question.
I have to study more MockClient internals to answer it :)

Actually, here sync-group request was not sent from the first call of `joinGroupIfNeeded` because of `Timer`.
It set to `autoTick=1` and timeouts right after the first call of `poll` see `ConsumerNetworkClient#poll` line 211.

I updated first timer to 2 so now both requests(join-group, sync-group) sent from the first call of `joinGroupIfNeeded`.
I added asserts for expected `generation` values after each `coordinator` call.

Hope, the test is clear now.

Reply:
Thanks for looking into this!! I tested it locally and it passes here.

Reply:
Also ran client:tests and all extended `BaseConsumerTest` and all passed.